File: tests/python/test_zero_order.py
from __future__ import annotations

import numpy as np

import optlib


def test_zero_order_methods_accept_python_value_function() -> None:
    target = np.array([1.5, -2.0])

    def value(values: np.ndarray) -> float:
        shifted = values - target
        return float(np.dot(shifted, shifted))

    for method in ["nelder_mead", "powell", "coordinate_search"]:
        result = optlib.MinimizeZeroOrder(
            value,
            np.array([-2.0, 3.0]),
            method=method,
            max_iter=400,
            step_tolerance=1e-7,
            function_tolerance=1e-12,
            initial_step=0.8,
            line_search_radius=3.0,
            log_trajectory=True,
        )
        assert result["value"] < 1e-6
        assert result["function_evaluations"] > 0
        assert result["trajectory"]["x"].shape[1] == 2


def test_zero_order_methods_reduce_rosenbrock() -> None:
    for method in ["nelder_mead", "powell", "coordinate_search"]:
        result = optlib.MinimizeRosenbrockZeroOrder(
            np.array([-1.2, 1.0]),
            method=method,
            max_iter=600,
            step_tolerance=1e-7,
            function_tolerance=1e-12,
            initial_step=0.8,
            line_search_radius=2.0,
            log_trajectory=False,
        )
        assert result["value"] < 1.0

File: tests/cpp/TestZeroOrder.cpp
#include <gtest/gtest.h>

#include <optlib/core/optimizers/ZeroOrder.hpp>

namespace {

double QuadraticValue(std::span<const double> values)
{
    auto sum = 0.0;
    for (double value : values) {
        auto shifted = value - 1.5;
        sum += shifted * shifted;
    }
    return sum;
}

optlib::ZeroOrderConfig ConfigFor()
{
    optlib::ZeroOrderConfig config;
    config.Criteria.MaxIterations = 300;
    config.Criteria.StepTolerance = 1e-7;
    config.Criteria.FunctionTolerance = 1e-12;
    config.InitialStep = 0.8;
    config.LineSearchRadius = 2.0;
    config.StoreTrajectory = true;
    return config;
}

} // namespace

TEST(ZeroOrderTest, MethodsConvergeOnQuadratic)
{
    const optlib::Vector initialX{-2.0, 4.0};
    for (auto method : {optlib::ZeroOrderMethod::NelderMead, optlib::ZeroOrderMethod::Powell,
                        optlib::ZeroOrderMethod::CoordinateSearch}) {
        const auto result = optlib::MinimizeZeroOrder(QuadraticValue, initialX.Span(), method, ConfigFor());
        EXPECT_LT(result.Value, 1e-6);
        EXPECT_GT(result.FunctionEvaluations, 0U);
    }
}

TEST(ZeroOrderTest, MethodsReduceRosenbrock)
{
    const optlib::Vector initialX{-1.2, 1.0};
    for (auto method : {optlib::ZeroOrderMethod::NelderMead, optlib::ZeroOrderMethod::Powell,
                        optlib::ZeroOrderMethod::CoordinateSearch}) {
        auto config = ConfigFor();
        config.Criteria.MaxIterations = 600;
        const auto result = optlib::MinimizeRosenbrockZeroOrder(initialX.Span(), method, config);
        EXPECT_LT(result.Value, 1.0);
    }
}
